File: packages/java/src/test/java/ai/alumnium/unit/accessibility/ChromiumAccessibilityTreeTest.java
package ai.alumnium.unit.accessibility;

import static org.assertj.core.api.Assertions.assertThat;

import ai.alumnium.accessibility.ChromiumAccessibilityTree;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class ChromiumAccessibilityTreeTest {

  @Test
  void rendersNodeValueAsAttribute() {
    Map<String, Object> node =
        Map.of(
            "nodeId", "1",
            "role", Map.of("value", "combobox"),
            "value", Map.of("value", "Option 2"));

    String xml = new ChromiumAccessibilityTree(Map.of("nodes", List.of(node))).toStr();

    assertThat(xml).contains("value=\"Option 2\"");
  }

  @Test
  void preservesUncheckedState() {
    Map<String, Object> node =
        Map.of(
            "nodeId",
            "1",
            "backendDOMNodeId",
            1,
            "role",
            Map.of("value", "checkbox"),
            "properties",
            List.of(Map.of("name", "checked", "value", Map.of("value", "false"))));
    ChromiumAccessibilityTree tree = new ChromiumAccessibilityTree(Map.of("nodes", List.of(node)));

    assertThat(tree.toStr()).contains("checked=\"false\"");
    assertThat(tree.scopeToArea(1).toStr()).contains("checked=\"false\"");
  }
}

File: packages/typescript/src/accessibility/ChromiumAccessibilityTree.test.ts
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { ChromiumAccessibilityTree } from "./ChromiumAccessibilityTree.ts";

const FIXTURE_PATH = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "__fixtures__/chromium_accessibility_tree.json",
);

describe("ChromiumAccessibilityTree", () => {
  it("renders node value as an attribute", () => {
    const tree = new ChromiumAccessibilityTree({
      nodes: [
        {
          nodeId: "1",
          role: { value: "combobox" },
          value: { value: "Option 2" },
        },
      ],
    });

    expect(tree.toStr()).toContain('value="Option 2"');
  });

  it("preserves unchecked state", () => {
    const tree = new ChromiumAccessibilityTree({
      nodes: [
        {
          nodeId: "1",
          role: { value: "checkbox" },
          properties: [
            { name: "invalid", value: { value: "false" } },
            { name: "checked", value: { value: "false" } },
          ],
        },
      ],
    });

    expect(tree.toStr()).toContain('checked="false"');
    expect(tree.toStr()).not.toContain("invalid");
  });

  describe("elementById", () => {
    it("returns correct element for given ID", async () => {
      const json = await fs.readFile(FIXTURE_PATH, "utf-8").then(JSON.parse);
      const tree = new ChromiumAccessibilityTree(json);
      expect(tree.elementById(1).backendNodeId).toBe(7);
      expect(tree.elementById(2).backendNodeId).toBe(6);
      expect(tree.elementById(3).backendNodeId).toBe(5);
    });
  });

  describe("scopeToArea", () => {
    it("returns the original tree when the element is not found", async () => {
      const json = await fs.readFile(FIXTURE_PATH, "utf-8").then(JSON.parse);
      const tree = new ChromiumAccessibilityTree(json);

      expect(tree.scopeToArea(99999).toStr()).toBe(tree.toStr());
    });

    it("preserves unchecked state in a scoped tree", () => {
      const tree = new ChromiumAccessibilityTree({
        nodes: [
          {
            nodeId: "1",
            backendDOMNodeId: 1,
            role: { value: "checkbox" },
            properties: [{ name: "checked", value: { value: "false" } }],
          },
        ],
      });

      expect(tree.scopeToArea(1).toStr()).toContain('checked="false"');
    });
  });
});
